### ISSUE TITLE
_disabled_torch_function_impl is unimplementable (must be special cased)

### ISSUE THREAD

**Issue by ezyang** [pytorch/pytorch#64690]
This is most easily seen in the following example, prior to fixing https://github.com/pytorch/pytorch/issues/64687

```
import torch
class A:
    @classmethod
    def __torch_function__(cls, func, types, args=(), kwargs=None):
        print(types)

a = A()
b = torch.nn.Parameter()

torch.add(a, b)
torch.overrides.handle_torch_function(None, (a, b), a, b)
```

this prints:

```
(<class '__main__.A'>,)
(<class '__main__.A'>, <class 'torch.nn.parameter.Parameter'>)
```

Parameter is completely omitted from the types in the first case, which goes through the C++ parser. The Python code path via buggy `handle_torch_function` shows what a naive implementation that didn't handle this case explicitly does: it includes Parameter in the types list, because, well, that's what it's supposed to do.

This is not very nice because although we originally motivated disabled_torch_function_impl as a pure performance optimization, it actually also has semantic meaning: classes who disable torch function impl don't show up in the types list. This is important because other implementations of torch function may query the types list to determine if they understand the arguments or not. 

Unfortunately, this behavior is also quite desirable. Canonically, we are supposed to be able to mark anything as a Parameter and expect everything to work as is with it. So if someone studiously goes through all the arguments and tests that they are, in fact, Tensors, would in fact fail if they were passed a Parameter. So having disabled_torch_function_impl work this way increases interoperability of said classes with other use cases (we are basically saying, "yeah, this is a subclass, but just pretend it is a Tensor for all intents and purposes). So we probably won't fix this issue, but I want to note it down as something to be aware of.

**Comment by ezyang** [comment]
So, based on #64693 (where people are super()'ing into Parameter), I feel like the right thing to do here is fix the fastpath to have semantic parity with `_disabled_torch_function_impl`, and then redo the semantics of `_disabled_torch_function_impl` to carefully defer to other Tensor objects for `__torch_function__` implementation in as many cases as possible (and make sure people writing other `__torch_function__` implementations don't reject Parameter).

**Comment by albanD** [comment]
Do I understand the problem correctly here that we want a very special handling of these subclasses in the sense that every other implementation should be tried before calling into this one? And we cannot do that today as we cannot make that function return NotImplemented all the time and we cannot force it to be a parent class for everything.

Would a more satisfying implementation be to "unpack" all the elements and call back into the given function (without disabling torch_function!)?

**Comment by ezyang** [comment]
> Do I understand the problem correctly here that we want a very special handling of these subclasses in the sense that every other implementation should be tried before calling into this one?

No, not quite. Right now, Parameter gets handled very specially, but this is an implementation accident. The question is whether or not it is *worth* having special handling for Parameter, or if we should force ourselves to only do things expressible with bare `__torch_function__`. Right now I'm leaning to the latter.

> Would a more satisfying implementation be to "unpack" all the elements and call back into the given function (without disabling torch_function!)?

This probably works but if we actually do the unpacks whenever there is a Parameter call this would seriously regress regular nn.Module performance.

**Comment by ezyang** [comment]
Here's what I think we should do.

```
// Where does _disabled_torch_function_impl come from?  It's main reason for
// existing is to make operations on Parameter fast: Parameter is a ubiquitous
// Tensor subclass, but it should get dropped whenever you do an operation on it
// (i.e., it should be treated as if it were just a Tensor).  This dropping
// should happen entirely from C++, so that operations on Parameter are just as
// fast as operations on Tensor (which bypass __torch_function__ entirely).
//
// Although Parameter is optimized in this way, we would still like to describe
// the semantics of _disabled_torch_function_impl as if it were implemented as
// any old __torch_function__.  This is what the function below does, but
// previously it was implemented incorrectly: in the old implementation, it
// simply disabled torch function and redispatched; this is wrong because
// Parameter may be involved in an operation with another custom tensor, and in
// that case we are obligated to attempt the __torch_function__ on that tensor
// first.  This was the cause of https://github.com/pytorch/pytorch/issues/64687
//
// In fact, the idea of just dropping Parameter from the types list is
// problematic.  Suppose we have two unrelated Tensor subclasses A and Parameter
// (which is _disabled_torch_function_impl).  Under the optimization, we ignore
// Parameter, and directly dispatch to A.__torch_function__ with types = (A,)
// (no Parameter in the list), no matter if I say A() + Parameter() or
// Parameter() + A().  Without the optimization, when I say A() + Parameter(),
// I'm obligated to call A.__torch_function__ first, with types = (A,
// Parameter), disagreeing with what you would see in the optimized case.
//
// The conclusion, then, is the following things must be true:
//
//  - Parameter *must* appear in the types list for other __torch_function__
//    calls, no matter which argument Parameter
//  - Parameter.__torch_function__ must only redispatch with torch function
//    disabled ONLY if there are no other nontrivial __torch_function__
//    implementations in the types list
//  - Other tensor subclasses must accept (not reject) occurrences of
//    Parameter in the types list
//
// So, putting it all together (here I use Parameter as shorthand for
// _disabled_torch_function_impl):
//
//  - When we collect overloaded arguments, Parameter counts as an overloaded
//    argument
//  - However, if there is another eligible overloaded type argument, Parameter
//    will always return NotImplemented to defer to it.  This means it's always
//    valid to skip Parameter when considering what overloaded functions to call
//  - If *everything* is Parameter, then we want to disable torch function and
//    redispatch to the underlying
//
// One subtlety is detecting if you are at the end of the chain.  Because types
// also doubles as the list of things we are dispatching, you simply have to
// test if your class is at the end of the chain.
```

**Comment by albanD** [comment]
Thanks for the write up I would have a couple comments:
- Do we want to implement this the way you described above? Or do we want a c++ fast-path (with proper semantic equivalentce) to avoid having to come back to python for every op that happen on only Parameter?
- Should we rename that function as "disabled" doesn't describe it very well.
- Actually this is a general use case (as seen in some tests that were added recently) where we do not want the automatic wrapping on exit that the default `__torch_function__` implementation does. Do we want to add another "default" torch function for that as well to allow the users to change that behavior easily?

**Comment by ezyang** [comment]
> Do we want to implement this the way you described above? Or do we want a c++ fast-path (with proper semantic equivalentce) to avoid having to come back to python for every op that happen on only Parameter?

Fast path is mandatory, Parameter is run a lot.

> Should we rename that function as "disabled" doesn't describe it very well.

Maybe? It's internal naming so whatever?
